metals/src/main/scala/scala/meta/internal/metals/ConnectionProvider.scala

```diff
@@ -637,56 +710,120 @@ class ConnectionProvider(
                         )
                         .asScala
+                        .withInterrupt
                         .flatMap {
                           case Messages.ImportProjectFailedSuggestBspSwitch.switchBsp =>
-                            switchBspServer()
-                          case _ => Future.unit
+                            switchBspServer().withInterrupt
+                          case _ => Interruptable.successful(BuildChange.Failed)
                         }
                     case _ =>
                       languageClient.showMessage(Messages.ImportProjectFailed)
+                      Interruptable.successful(BuildChange.Failed)
                   }
-                  Future.successful(BuildChange.Failed)
                 }
             } yield change
-          } else Future.successful(BuildChange.None)
+          } else Interruptable.successful(BuildChange.None)
         }
       } yield change
     }
   }
 
-  def switchBspServer(): Future[Unit] =
+  def switchBspServer(): Future[BuildChange] =
     withWillGenerateBspConfig {
       for {
         connectKind <- bspConnector.switchBuildServer()
-        _ <-
+        change <-
           connectKind match {
-            case None => Future.unit
+            case None => Future.successful(BuildChange.Failed)
             case Some(SlowConnect) =>
               slowConnectToBuildServer(forceImport = true)
             case Some(request: ConnectRequest) => connect(request)
           }
-      } yield ()
+      } yield change
     }
 }
 
 sealed trait ConnectKind
 object SlowConnect extends ConnectKind
 
-sealed trait ConnectRequest extends ConnectKind
+sealed trait ConflictBehaviour
+case object Yield extends ConflictBehaviour
+case object TakeOver extends ConflictBehaviour
+case object Queue extends ConflictBehaviour
+
+sealed trait ConnectRequest extends ConnectKind {
+
+  /**
+   * Decides what to do with a new connect request
+   * in presence of an another ongoing/queued request.
+   * @param other the ongoing or queued request
+   * @return behavoiur of the incoming request
+   * Yield    -- cancel this
+   * TakeOver -- cancel other
+   * Queue    -- queue
+   */
+  def cancelCompare(other: ConnectRequest): ConflictBehaviour
+}
 
-case class Disconnect(shutdownBuildServer: Boolean) extends ConnectRequest
-case class Index(check: () => Unit) extends ConnectRequest
-case class ImportBuildAndIndex(bspSession: BspSession) extends ConnectRequest
-case class ConnectToSession(bspSession: BspSession) extends ConnectRequest
+case class Disconnect(shutdownBuildServer: Boolean) extends ConnectRequest {
+  def cancelCompare(other: ConnectRequest): ConflictBehaviour =
+    other match {
+      case _: Index => Queue
+      case _ => Yield
+    }
+}
+case class Index(check: () => Unit) extends ConnectRequest {
+  def cancelCompare(other: ConnectRequest): ConflictBehaviour =
+    other match {
+      case _: Disconnect => Queue
```

Comment:
If we have a Disconnect then we should probably Yield, no? There is no sense to run idnexing after disconnecting?